samza-core/src/main/java/org/apache/samza/task/AsyncRunLoop.java

```diff
@@ -597,7 +599,9 @@ private boolean checkEndOfStream() {
         if (envelope.isEndOfStream()) {
           SystemStreamPartition ssp = envelope.getSystemStreamPartition();
           processingSspSet.remove(ssp);
-          pendingEnvelopeQueue.remove();
+          if (!hasIntermediateStreams) {
```

Comment:
So, if the task has intermediate streams, pendingEnvelopeQueue will not be empty. How do we trigger EoS in this case?